src/Components/Components/src/RenderTree/Renderer.cs

```diff
@@ -111,13 +118,49 @@ private static IComponentActivator GetComponentActivatorOrDefault(IServiceProvid
         /// </summary>
         protected internal ElementReferenceContext? ElementReferenceContext { get; protected set; }
 
+        internal HotReloadContext HotReloadContext { get; } = new();
+
+        private async void RenderRootComponentsOnHotReload()
+        {
+            await Dispatcher.InvokeAsync(async () =>
+            {
+                Debug.Assert(_rootComponents is not null);
+                HotReloadContext.IsHotReloading = true;
+
+                try
+                {
+                    _pendingTasks = new List<Task>();
+                    foreach (var (componentState, initialParameters) in _rootComponents)
+                    {
+                        AddToPendingTasks(componentState.Component.SetParametersAsync(initialParameters));
+                    }
+
+                    await ProcessAsynchronousWork();
```

Comment:
Are you sure we should really be doing this asynchronously? That could take a completely arbitrary amount of time, maybe minutes (waiting for any component in the hierarchy to load data). During this indefinite period, there could be:

 * Arbitrary user interactions, which will render weirdly because of the hot reload flag
 * Other hot reload deltas being applied, interfering with the shared `_pendingTasks` field

I'm suspicious that this should really only be synchronous. We only need the component tree to re-render itself once immediately to update the state in the existing UI. If components do some async data loading, it's up to them to be able to re-render themselves afterwards.